Add --disable-* options for SIMD instruction sets and other configure flags

This patch adds AC_ARG_ENABLE options for maintainer-mode, openmp, and various
SIMD instruction sets (SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2, FMA, FMA4, AVX,
AVX2, AVX512F, AVX512DQ, AVX512VL) to linbox configure.ac. When these options
are disabled, the corresponding -mno-* compiler flags are added to OPTIM_FLAGS.

--- a/configure.ac
+++ b/configure.ac
@@ -76,6 +76,10 @@
 
 AS_ECHO([---------------------------------------])
 
+# Check for compatibility options (maintainer-mode, openmp handled elsewhere)
+LB_OPT_MAINTAINER_MODE
+LB_OPT_OPENMP
+
 # Set OPTIM_FLAGS, DEBUG_FLAGS depending on compiler and command line arguments
 SET_FLAGS
 
--- a/macros/common.m4
+++ b/macros/common.m4
@@ -98,14 +98,67 @@
 AS_VAR_POPDEF([CACHEVAR])dnl
 ])dnl AX_CHECK_COMPILE_FLAGS
 
+dnl Check for --disable-maintainer-mode option (ignored, for compatibility)
+AC_DEFUN([LB_OPT_MAINTAINER_MODE],[
+    AC_ARG_ENABLE(maintainer-mode,
+        [AC_HELP_STRING([--disable-maintainer-mode],
+            [disable maintainer mode (ignored, for compatibility)])],
+        [], [])
+])
+
+dnl Check for --disable-openmp option (ignored here, handled elsewhere)
+AC_DEFUN([LB_OPT_OPENMP],[
+    AC_ARG_ENABLE(openmp,
+        [AC_HELP_STRING([--disable-openmp],
+            [disable OpenMP (ignored here, for compatibility)])],
+        [], [])
+])
+
+dnl Macro to add -mno-* flag to OPTIM_FLAGS if --disable-* is passed
+dnl Usage: LB_OPT_SIMD_FLAG(name, flag)
+dnl   name: the option name (e.g., sse3)
+dnl   flag: the compiler flag (e.g., -mno-sse3)
+AC_DEFUN([LB_OPT_SIMD_FLAG],[
+    AC_ARG_ENABLE($1,
+        [AC_HELP_STRING([--disable-$1],
+            [add $2 to compiler flags])],
+        [AS_IF([test "x$enableval" = "xno"],
+            [AX_CHECK_COMPILE_FLAG([$2],
+                [AC_MSG_NOTICE([Adding '$2' to OPTIM_FLAGS])
+                 OPTIM_FLAGS+=" $2"],
+                [AC_MSG_NOTICE([Compiler does not support $2, ignoring --disable-$1])])])],
+        [])
+])
+
+dnl Check for all SIMD disable options and add corresponding -mno-* flags
+AC_DEFUN([LB_OPT_SIMD],[
+    dnl SSE options
+    LB_OPT_SIMD_FLAG([sse], [-mno-sse])
+    LB_OPT_SIMD_FLAG([sse2], [-mno-sse2])
+    LB_OPT_SIMD_FLAG([sse3], [-mno-sse3])
+    LB_OPT_SIMD_FLAG([ssse3], [-mno-ssse3])
+    LB_OPT_SIMD_FLAG([sse41], [-mno-sse4.1])
+    LB_OPT_SIMD_FLAG([sse42], [-mno-sse4.2])
+
+    dnl FMA options
+    LB_OPT_SIMD_FLAG([fma], [-mno-fma])
+    LB_OPT_SIMD_FLAG([fma4], [-mno-fma4])
+
+    dnl AVX options
+    LB_OPT_SIMD_FLAG([avx], [-mno-avx])
+    LB_OPT_SIMD_FLAG([avx2], [-mno-avx2])
+    LB_OPT_SIMD_FLAG([avx512f], [-mno-avx512f])
+    LB_OPT_SIMD_FLAG([avx512dq], [-mno-avx512dq])
+    LB_OPT_SIMD_FLAG([avx512vl], [-mno-avx512vl])
+])
 
 dnl Append -march=native or -mcpu=native (if recognized by the compiler) to
 dnl OPTIM_FLAGS if not present in CXXFLAGS and not cross-compiling and
 dnl --without-archnative is not set
 AC_DEFUN([ARCH_FLAGS],[
     AC_ARG_WITH(archnative, [AC_HELP_STRING([--without-archnative],
-        [do not use -march=native or -mcpu=native (default is to use it if not already present in CXXFLAGS)])])
-
+        [do not use -march=native or -mcpu=native (default is to use it if not already present in CXXFLAGS)])
+])
     AX_CHECK_COMPILE_FLAG([-march=native], [
         AS_CASE([$CXXFLAGS],
                 [*-march=*], [], # do nothing if already set in CXXFLAGS
@@ -127,6 +180,10 @@
                             OPTIM_FLAGS+=" -mcpu=native"])])])
         ], [], [], [])
     ], [], [])
+
+    dnl Process SIMD disable options after setting arch flags
+    dnl This allows --disable-avx512f etc. to add -mno-* flags
+    LB_OPT_SIMD
 ])
 
 dnl Append -mfpmath=sse to OPTIM_FLAGS on i386 and i686 architecture with SSE
